src/Processors/Transforms/MergeSortingTransform.cpp

```diff
@@ -166,7 +186,33 @@ void MergeSortingTransform::consume(Chunk chunk)
       *  will merge blocks that we have in memory at this moment and write merged stream to temporary (compressed) file.
       * NOTE. It's possible to check free space in filesystem.
       */
-    if (max_bytes_before_external_sort && sum_bytes_in_blocks > max_bytes_before_external_sort)
+    bool external_sort = max_bytes_before_external_sort && sum_bytes_in_blocks > max_bytes_before_external_sort;
+    if (!external_sort && max_bytes_ratio_before_external_sort > 0.)
+    {
+        auto [current_memory_usage, memory_limit] = getCurrentQueryMemoryUsageAndLimit();
+        if (current_memory_usage > memory_limit * max_bytes_ratio_before_external_sort)
```

Comment:
here probably is better to use `sum_bytes_in_blocks`

Reply:
Hm, but it does not takes into account memory usage in other places, i.e. aggregation